# Data Manipulation tasks:
    
- Filtering data

- Pivot Table

In [1]:
#store.csv
import pandas as pd
store=pd.read_csv("D://Files//Store.csv")

In [2]:
#view first 5 observation
store.head()

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
2           2       3  CA-2013-138688   6/13/2013   6/17/2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   

  Customer ID    Customer Name    Segment        Country  ... Postal Code  \
0    CG-12520      Claire Gute   Consumer  United States  ...       42420   
1    CG-12520      Claire Gute   Consumer  United States  ...       42420   
2    DV-13045  Darrin Van Huff  Corporate  United States  ...       90036   
3    SO-20335   Sean O'Donnell   Consumer  United States  ...       33311   
4    SO-20335   Sean O'Donnell   Consumer  United States  ...       33311   

  Region       Product ID         Category Sub-Category  \
0  South  FUR-BO-10001798        Furniture    Bookcases   
1  South  FUR-CH-10000454        Furniture       Chairs   
2   West  OFF-LA-10000240  Office Supplies       Labels   
3  South  FUR-TA-10000577        Furniture       Tables   
4  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 22 columns]

In [4]:
# number of observations
store.shape[0]

9994

In [5]:
# information about each variable
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     9994 non-null   int64  
 1   Row ID         9994 non-null   int64  
 2   Order ID       9994 non-null   object 
 3   Order Date     9994 non-null   object 
 4   Ship Date      9994 non-null   object 
 5   Ship Mode      9994 non-null   object 
 6   Customer ID    9994 non-null   object 
 7   Customer Name  9994 non-null   object 
 8   Segment        9994 non-null   object 
 9   Country        9994 non-null   object 
 10  City           9994 non-null   object 
 11  State          9994 non-null   object 
 12  Postal Code    9994 non-null   int64  
 13  Region         9994 non-null   object 
 14  Product ID     9994 non-null   object 
 15  Category       9994 non-null   object 
 16  Sub-Category   9994 non-null   object 
 17  Product Name   9994 non-null   object 
 18  Sales   

Orderdate and Ship date --> data types should be changed into datetime data type.

In [7]:
#summary statistics for continuous variables
store.describe()

Unnamed: 0       Row ID   Postal Code         Sales     Quantity  \
count  9994.000000  9994.000000   9994.000000   9994.000000  9994.000000   
mean   4996.500000  4997.500000  55190.379428    229.858001     3.789574   
std    2885.163629  2885.163629  32063.693350    623.245101     2.225110   
min       0.000000     1.000000   1040.000000      0.444000     1.000000   
25%    2498.250000  2499.250000  23223.000000     17.280000     2.000000   
50%    4996.500000  4997.500000  56430.500000     54.490000     3.000000   
75%    7494.750000  7495.750000  90008.000000    209.940000     5.000000   
max    9993.000000  9994.000000  99301.000000  22638.480000    14.000000   

          Discount       Profit  
count  9994.000000  9994.000000  
mean      0.156203    28.656896  
std       0.206452   234.260108  
min       0.000000 -6599.978000  
25%       0.000000     1.728750  
50%       0.200000     8.666500  
75%       0.200000    29.364000  
max       0.800000  8399.976000

- average sales is 229.9
- average profit is 28.66
- average quantity sold 3.79
- mean >median  in sales,Profit--> Sales and Profit variable is highly positively skewed

In [9]:
store.describe(include='O')

Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
count             9994        9994        9994            9994        9994   
unique            5009        1238        1334               4         793   
top     CA-2014-100111  09-06-2013  12/16/2012  Standard Class    WB-21850   
freq                14          38          35            5968          37   

        Customer Name   Segment        Country           City       State  \
count            9994      9994           9994           9994        9994   
unique            793         3              1            531          49   
top     William Brown  Consumer  United States  New York City  California   
freq               37      5191           9994            915        2001   

       Region       Product ID         Category Sub-Category     Product Name  
count    9994             9994             9994         9994             9994  
unique      4             1862                3           17             1850  
top      West  OFF-PA-10001970  Office Supplies      Binders  Staple envelope  
freq     3203               19             6026         1523               48

In [8]:
# How many cities are the orders being delivered to?
len(store.City.unique())

531

#### Filter

In [ ]:
# logical indexation : data[logical condition]
# query method: data.query('logical condition')

In [10]:
store['Segment'].unique()

array(['Consumer', 'Corporate', 'Home Office'], dtype=object)

In [12]:
# filter the data by segments are consumer
#logical indexation
store1=store[store['Segment']=='Consumer']
store1.tail(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
9992        9992    9993  CA-2014-121258   2/27/2014  03-04-2014   
9993        9993    9994  CA-2014-119914  05-05-2014  05-10-2014   

           Ship Mode Customer ID Customer Name   Segment        Country  ...  \
9992  Standard Class    DB-13060   Dave Brooks  Consumer  United States  ...   
9993    Second Class    CC-12220  Chris Cortes  Consumer  United States  ...   

     Postal Code Region       Product ID         Category Sub-Category  \
9992       92627   West  OFF-PA-10004041  Office Supplies        Paper   
9993       92683   West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name   Sales Quantity  \
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.60        4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.16        2   

      Discount  Profit  
9992       0.0  13.320  
9993       0.0  72.948  

[2 rows x 22 columns]

In [13]:
store1.shape[0]

5191

In [ ]:
#query method
store2=store.query('Segment=="Consumer"')
store2.head(2)

In [14]:
#query
store.query('Segment=="Consumer"').shape[0]

5191

In [ ]:
#How many orders were placed for region south and/or for segment consumer
store[(store['Region']=='South') & (store['Segment']=='Consumer')].shape[0]

In [ ]:
#query method
store.query('Region=="South" & Segment=="Consumer"').shape[0]

In [ ]:
#or condition
store.query('Region=="South" | Segment=="Consumer"').shape[0]

In [ ]:
#logical indexation
store[(store["Region"]=="South") | (store["Segment"]=="Consumer")].shape[0]

In [15]:
# Find out total quantity of products shipped to the east region?
store.query('Region=="East"')['Quantity'].sum()

10618

In [18]:
store.query('Region=="East"')[['Quantity','Segment']]

Quantity   Segment
23           2  Consumer
27           7  Consumer
28           2  Consumer
29           3  Consumer
30           2  Consumer
...        ...       ...
9968         2  Consumer
9969         7  Consumer
9981         5  Consumer
9984        10  Consumer
9985         4  Consumer

[2848 rows x 2 columns]

In [19]:
store1.to_csv("D://Files//filtereddata.csv")

### Pivot table
pandas.pivot_table(index, columns, values) function produces pivot table based on 3 columns of the DataFrame.

Uses unique values from index / columns and fills with values.

Python Pandas function pivot_table help us with the summarization and conversion of dataframe in long form to dataframe in wide form, in a variety of complex scenarios.

In [20]:
df = pd.DataFrame({'First Name': ['Aryan', 'Rohan', 'Riya', 'Yash', 'Siddhant', ],
                   'Last Name': ['Singh', 'Agarwal', 'Shah', 'Bhatia', 'Khanna'],
                   'Type': ['Full-time Employee', 'Intern', 'Full-time Employee', 
                            'Part-time Employee', 'Full-time Employee'],
                   'Department': ['Administration', 'Technical', 'Administration', 
                                  'Technical', 'Management'],
                   'YoE': [2, 3, 5, 7, 6],
                   'Salary': [20000, 5000, 10000, 10000, 20000]})

df

First Name Last Name                Type      Department  YoE  Salary
0      Aryan     Singh  Full-time Employee  Administration    2   20000
1      Rohan   Agarwal              Intern       Technical    3    5000
2       Riya      Shah  Full-time Employee  Administration    5   10000
3       Yash    Bhatia  Part-time Employee       Technical    7   10000
4   Siddhant    Khanna  Full-time Employee      Management    6   20000

Use pd.pivot_table and specify the data, index, columns, aggfunc and `values` parameters.

In [21]:
# create pivot table of average salary for employees based on their Department and Job type
pd.pivot_table(index='Department',columns="Type",values='Salary',data=df)

Type            Full-time Employee  Intern  Part-time Employee
Department                                                    
Administration             15000.0     NaN                 NaN
Management                 20000.0     NaN                 NaN
Technical                      NaN  5000.0             10000.0

More use cases of the pivot_table function could be used

https://pbpython.com/pandas-pivot-table-explained.html



In [23]:
###filter()
l=[6,7,8,34,56]
# create a list of even numbers
#filter(fun,iterable)
list(filter(lambda x:x%2==0,l))

[6, 8, 34, 56]

In [25]:
#map()
y=lambda x: x%2==0
#map(fun,iterable)
list(map(y,l))

[True, False, True, True, True]